v2/car.go

```diff
@@ -43,6 +43,9 @@ type (
 	}
 )
 
+// fullyIndexedCharPos is the position of Characteristics.Hi bit that specifies whether the index is a catalog af all CIDs or not.
+const fullyIndexedCharPos = 0
```

Comment:
this will largely depend on the spec, but if you're using the lowest bit of the high uint64, then you're assigning one of the two bits right in the middle :)

Reply:
another way to verify this is to e.g. hex encode the characteristic bits, and see that they visually correspond to what you'd expect - e.g. the "fully indexed" bit being the first, thus all the way at the left (highest)

Reply:
I *might* be wrong on all this, because thinking about bits and endianess is sometimes tricky. All the more reason to visualize how the bits are laid out when encoded via a test.

Reply:
> tricky

An understatement, trying to pin this all down for the HAMT spec was crazy and we've still ended up with differing implementations and the cross-language challenges are a headache.

Arguably using `Hi` is the correct choice here if we're thinking of this as an "ordered" bit field since we're writing it out first. I think even this didn't need to even be a pair of integers but just a `[16]byte` that we manipulate directly (just not with `big.Int` _please!).

I wrote https://github.com/rvagg/bit-sequence to do JS & Go versions of what it means to slice a bit field into sections of integers by arbitrary bit boundaries. It's not quite the same thing here since we're dealing with a single bit, but maybe we could verify that what's being set here is a 1-bit integer at the position in question, and then we'd have a consistent reference to point to at least.

Not something to get too hung up on, however. There's a first-mover advantage here in that what you define now is going to be locked in for any other implementation and they'll (including JS) will just have to roll with that.

Reply:
+1 to making characteristics a byte array, actually. If you remove the complexity of having to think about endianness, I think our code and docs would get a lot simpler.

The spec allows it - it only says it's 128 bits or 16 bytes, it doesn't require any representation in memory.

Technically a breaking change in terms of Go APIs, but I think we're in the clear given how the characteristics bits are entirely unused right now.

Reply:
> maybe we could verify that what's being set here is a 1-bit integer at the position in question, and then we'd have a consistent reference to point to at least.

Does [this test](https://github.com/ipld/go-car/blob/9f9f687b3606dae462007d93cc0e8684129435f5/v2/car_test.go#L206-L221) do the trick on verification?


> +1 to making characteristics a byte array, actually.

I think it was a byte array at first but we settled on `Hi` `Low` int64; I am failing to find reference discussions apologies. 
Happy to change it in a separate PR if one of you gents would kindly volunteer to write up an issue?

Reply:
Thanks for capturing this @mvdan https://github.com/ipld/go-car/issues/240

PR updated to use the left most bit in `Hi`